viewer/src/Main.rs

```diff
@@ -0,0 +1,324 @@
+    }
+
+    fn draw_text(
+        canvas: &mut WindowCanvas,
+        texture_creator: &TextureCreator<WindowContext>,
+        font: &sdl2::ttf::Font,
+        x: i32,
+        y: i32,
+        tag: i32,
+        len: i32,
+    ) -> Result<(), String> {
+        let tag_text: String = format!("{0},{1}", tag.to_string(), len.to_string());
+        let surface = font
+            .render(&tag_text)
+            .blended(Color::RGBA(0, 0, 0, 128))
+            .map_err(|e| e.to_string())?;
+
+        let texture = texture_creator
+            .create_texture_from_surface(&surface)
+            .map_err(|e| e.to_string())?;
+
+        canvas.set_draw_color(Color::RGB(255, 255, 255));
+        canvas.fill_rect(Rect::new(x, y, (tag_text.len() * 20) as u32, 50 as u32))?;
+
+        let target = Rect::new(x, y, (tag_text.len() * 20) as u32, 50 as u32);
+        canvas.copy(&texture, None, Some(target))?;
+
+        Ok(())
+    }
+
+    pub fn run<'a>(&mut self, spans: Vec<Span>, font: Font<'a, 'static>) -> Result<(), String> {
+        let mut event_pump = self.sdl_context.event_pump()?;
+        let mut prev_keycount: i32 = 0;
+        let mut keycount: i32 = 0;
+        let mut draw_x = 0;
+        let mut draw_y = 0;
+        let mut draw_tag = 0;
+        let mut draw_len = 0;
+
+        'running: loop {
+            for event in event_pump.poll_iter() {
+                match event {
+                    Event::Quit { .. } => break 'running,
+                    Event::KeyDown {
+                        keycode: Some(keycode),
+                        ..
+                    } => {
+                        keycount += 1;
+                        match keycode {
+                            Keycode::Q => {
+                                self.scale = self
+                                    .scale
+                                    .saturating_add((keycount * keycount).try_into().unwrap())
+                            } //plus
+                            Keycode::W => {
+                                self.scale = self
+                                    .scale
+                                    .saturating_sub((keycount * keycount).try_into().unwrap())
+                            } //minus
+                            Keycode::E => {
+                                self.scale =
+                                    (self.max_stop - self.min_start) / (self.window_width as u64)
+                            } //reset
+                            Keycode::A => self.scroll = self.scroll.saturating_add(keycount),
+                            Keycode::S => self.scroll = self.scroll.saturating_sub(keycount),
+                            Keycode::D => self.scroll = 0,
+                            _ => {}
+                        }
+                    }
+                    Event::MouseButtonDown { x, y, .. } => {
+                        for zone in &self.draw_zones {
+                            if zone.x_start < x.try_into().unwrap()
```

Comment:
I think some of these try_intos may be unnecessary, see what `cargo clippy` reports for this project